presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java

```diff
@@ -0,0 +1,275 @@
+
+        return new QuerySpecification(
+                rewrittenSelect,
+                rewrittenFrom,
+                rewrittenWhere,
+                rewrittenGroupBy,
+                rewrittenHaving,
+                rewrittenOrderBy,
+                node.getLimit());
+    }
+
+    @Override
+    protected Node visitSelect(Select node, Void context)
+    {
+        if (!isRewritable) {
+            return node;
+        }
+        ImmutableList.Builder<SelectItem> rewrittenSelectItems = ImmutableList.builder();
+
+        for (SelectItem selectItem : node.getSelectItems()) {
+            SelectItem rewrittenSelectItem = (SelectItem) process(selectItem, context);
+            rewrittenSelectItems.add(rewrittenSelectItem);
+        }
+
+        return new Select(node.isDistinct(), rewrittenSelectItems.build());
+    }
+
+    @Override
+    protected Node visitSingleColumn(SingleColumn node, Void context)
+    {
+        return new SingleColumn((Expression) process(node.getExpression(), context), node.getAlias());
+    }
+
+    @Override
+    protected Node visitArithmeticBinary(ArithmeticBinaryExpression node, Void context)
+    {
+        return new ArithmeticBinaryExpression(
+                node.getOperator(),
+                (Expression) process(node.getLeft(), context),
+                (Expression) process(node.getRight(), context));
+    }
+
+    @Override
+    protected Node visitIdentifier(Identifier node, Void context)
+    {
+        if (!materializedViewInfo.getBaseToViewColumnMap().containsKey(node)) {
+            isRewritable = false;
+            return node;
+        }
+        return new Identifier(materializedViewInfo.getBaseToViewColumnMap().get(node).getValue(), node.isDelimited());
+    }
+
+    @Override
+    protected Node visitFunctionCall(FunctionCall node, Void context)
+    {
+        ImmutableList.Builder<Expression> rewrittenArguments = ImmutableList.builder();
+
+        if (materializedViewInfo.getBaseToViewColumnMap().containsKey(node)) {
+            Expression derivedExpression = materializedViewInfo.getBaseToViewColumnMap().get(node);
+            rewrittenArguments.add(derivedExpression);
+        }
+        else {
+            for (Expression argument : node.getArguments()) {
+                Expression rewrittenArgument = (Expression) process(argument, context);
+                rewrittenArguments.add(rewrittenArgument);
+            }
+        }
+
+        return new FunctionCall(
+                node.getName(),
+                node.getWindow(),
+                node.getFilter(),
+                node.getOrderBy(),
+                node.isDistinct(),
+                node.isIgnoreNulls(),
+                rewrittenArguments.build());
+    }
+
+    @Override
+    protected Node visitRelation(Relation node, Void context)
+    {
+        if (materializedViewInfo.getMaterializedViewDefinitionRelation().isPresent() && node.equals(materializedViewInfo.getMaterializedViewDefinitionRelation().get())) {
```

Comment:
Does this only compares the table name? How do we reject the following case?
1. select a different ds partition of the same table?
2. the table name doesn't match but the the table alias matches.

Reply:
@yuanzhanhku Sorry for not mentioning this in the code. I did not plan to handle cases like table alias (select b1.a FROM baseTable b1) in this version of optimizer. SInce we are not handling any JOIN or sub-query cases, I thought the cases with Table alias would become trivial. This could be handled in future PR (especially when dealing with multiple tables or self join).

But if you believe that this part should be implemented at this stage, I would work around handling those cases.

Reply:
It is fine to not handle it in this PR. But if there is way to easily reject the MV rewrite when the table alias exists, we'd better add it. Otherwise, adding a TODO is fine.

Besides, I think it is better to add the test cases even if the feature is not implemented. We can verify that the rewrite is not applied, it doesn't generate an invalid plan and it doesn't crash.

Reply:
@yuanzhanhku Sounds good. I'll updated my code shortly.